ladder-step1

clown0425 opened the pull request:
step1 제출합니다

clown0425 committed:
제출

godrm commented:
우선 코드리뷰에 앞서서 git commit 로그를 작성하는 방식, commit 하는 작업 단위를 어느 정도로 구분해야 하는지 학습해보세요.
한 번 PR에 커밋 하나만 보내면 안됩니다. 커밋은 변경하는 최소 작업단위로 작게 나눠야 합니다. 그럴려면 프로그램 구조도 작게 나눠야 합니다.
그래야 원하는 커밋만 선택할 수 있습니다. 이 부분을 우선적으로 정리해보세요.
커밋은 숙제검사가 아닙니다. 개발자가 요구사항을 만족하는 코드를 만들기 위해 어떤 작업을 했는지 이력이 그대로 보여야 합니다.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -0,0 +1,69 @@
+
+import Foundation
+
+let humNum = 10, ladderNum = 16, hight = 5
```
전역 변수를 사용하지 마세요. 
함수는 명확하게 입력과 출력이 연관이 있도록 순수하게 작성하는 게 좋습니다. 
스위프트 API 디자인가이드 문서를 읽고 변수 이름도 개선해보세요.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -0,0 +1,69 @@
+func len (humNum:Int)->Int{
+    let len = ((humNum-1) * 2) + 1
+    return len
+}
```
전역 변수와 같은 이름으로 매개 변수를 사용하게 있네요. 좋지 않은 습관입니다. 
변수 이름에 Num 이나 Arr, String 같은 타입 정보를 붙이지 말고 의도하는 역할을 명확하게 나타내는 단어를 선택하세요.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -0,0 +1,69 @@
+func emptyladderArr(len:Int, hight:Int)->[[Character]]{
+    let ladderArr : [[Character]] = Array(repeating: Array(repeating: " ",count:len ), count: hight)
+    return ladderArr
+}
```
스위프트 API 디자인 가이드 문서에서 객체를 생성하는 함수 이름에 대한 규칙을 찾아보세요.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+    return ladderArr
+}
```
사소한 것으로 생각할 수 있지만 늘 철자에 유의하세요.`height` 를 쓰려던 게 맞죠?
의도한 것과 다른 의미로 전달될 수 있습니다.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -0,0 +1,69 @@
+func ranWidth (humNum:Int)->Int{
+    let ranWidth = (Int(arc4random_uniform(UInt32(humNum-1))) * 2) + 1
+    return ranWidth
+}
```
여기 두 함수는 `run`도 아니고 `ran`을 사용하고 있네요. 어떤 의도에서 선택한 것인지 궁금합니다.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -0,0 +1,69 @@
+var ladderArr = emptyladderArr(len:lenValue, hight:hight)
+ladderArr = addHight(humNum:humNum,hight:hight,ladderArr:ladderArr)
+ladderArr = addWidth(ladderNum:ladderNum,hight:hight,humNum:humNum,ladderArr:ladderArr)
+resultprint(ladderArr: ladderArr)
```
메인 프로그램을 구성하는 것도 함수로 한 번 감싸서 표현해보세요. 
제가 얘기했던 프로그램 구성 요소를 만족하고 있는지 다시 한 번 확인해보세요.

godrm reviewed (commented):
다음과 같은 피드백을 드립니다. 꼼꼼하게 살펴보시고 개선해보세요.

clown0425 committed:
매개변수 이름과 전역변수 이름이 겹치지 않게 수정
변수이름을 좀더 직관적이고 일관되게 변경했습니다
배열은 arr로 표현하지 않고 복수형으로 표현
height로 철자 올바르게 고쳤습니다
ran은 random을 줄인 표현 이었습니다
이것도 줄이지 않고 정확하게 표현했습니다

clown0425 committed:
입력을 받는게 아니라 제가 직접 입력해서 작동하더 코드를
입력을 받는 코드로 바꾸고
메인 프로그램을 구성하는 것도 함수로 감싸 보았습니다.

clown0425 committed:
요구 사항을 다시 점검해 봤더니
가로 사다리 개수는 입력 받는게 아니어서
플레이어수와 높이를 바탕으로 가능한 개수중 랜덤으로 숫자를 생성해 배치하는 방향으로 설계해 보았습니다

clown0425 reviewed (commented):
변수 이름 수정

clown0425 reviewed (commented):
직접 입력하지 않고 입력을 받는 것으로 고쳤습니다

clown0425 reviewed (commented):
가로 사다리 개수를 입력 받는게 아니라 랜덤으로 생성하는 것으로 바꾸었습니다

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -7,63 +7,88 @@
+    let userResponse = readLine()
+    guard let number = Int(userResponse!) else{
+        fatalError("양의정수를 입력하세요")
+    }
```
fatalError()를 출력하기 보다는 항상 다른 방법을 찾아보세요.
여기서는 함수를 빠져나가서 상위 모듈이 무언가 에러 문구를 출력하거나 다시 호출할 수 있도록 하는게 좋습니다. 
다음 단계를 하면서 개선해보세요.

godrm commented on laddergame01/laddergame01/main.swift:
```diff
@@ -7,63 +7,88 @@
+    var ladders = emptyladders(width:ladderWidthLength, height:ladderHeight)
+    ladders = addHeight(width:playernumber,height:ladderHeight,ladders:ladders)
+    ladders = addWidth(ladderNumber:widthLadderNumber,height:ladderHeight,width:playernumber,ladders:ladders)
+    resultprint(ladders: ladders)
```
다음 단계를 진행하면서 여기 playLadderGame() 함수 부분을 이렇게 개선해보세요.
- 입력 / 처리 / 출력을 담당하는 하위 함수로 나눠져 있는가
- 내부에서 호출하는 하위 함수의 추상화 수준이 일정한가. 
- 하위 함수의 추상화 수준이란 무엇을 의미하는가
- 상위 함수는 간결하고 하위 함수가 더 구체적인 동작을 처리하고 있는가

godrm commented:
마지막 댓글을 달고 머지하려고 했는데 놓쳤던 것 같습니다.
머지할테니 pull 받아서 다음 단계를 진행하세요.
